### ISSUE TITLE
[Particles] ParticleBuffer -> Memory management

### ISSUE THREAD

**Issue by luisfb** [craftworkgames/MonoGame.Extended#580]
I have a question about the Dispose method of ParticleBuffer:

Inside the constructor it allocates the value of SizeInBytes property as follows:
`GC.AddMemoryPressure(SizeInBytes);`
The `SizeInBytes` returns `Particle.SizeInBytes*(Size + 1).`
However, inside the Dispose method, it calls:
`GC.RemoveMemoryPressure(Particle.SizeInBytes*Size);`
Shouldn't it be:
`GC.RemoveMemoryPressure(Particle.SizeInBytes*(Size + 1));`
?
I know the +1 is added for margin between head and tail, but I want to be sure the value being unallocated is right.
Thank you.

**Comment by craftworkgames** [comment]
Hey @luisfb 

First of all, let me just say nicely spotted. There certainly seems to be a bug in that code of some sort.

Secondly, I didn't actually write that code myself. The particle stuff was originally ported (with permission) from [mercury particle engine](http://matthew-davey.github.io/mercury-particle-engine/). 
>
If we take a look at the [original code](https://github.com/Matthew-Davey/mercury-particle-engine/blob/develop/Mercury.ParticleEngine.Core/ParticleBuffer.cs) it does appear to have the same problem (assuming we are correct about it being a bug).
>
Now, in terms of fixing the issue, it's probably best to take a look at the documentation for [`AddMemoryPressure`](https://docs.microsoft.com/en-us/dotnet/api/system.gc.addmemorypressure?view=netcore-2.2).
>
> In determining when to schedule garbage collection, the runtime takes into account how much managed memory is allocated. If a small managed object allocates a large amount of unmanaged memory, the runtime takes into account only the managed memory, and thus underestimates the urgency of scheduling garbage collection. The AddMemoryPressure method informs the runtime of this additional pressure on system memory.
>
> In the simplest usage pattern, a managed object allocates unmanaged memory in the constructor and releases it in the Dispose or Finalize method. Call the AddMemoryPressure method after allocating the unmanaged memory, and call the RemoveMemoryPressure method after releasing it.
>
> In more complicated scenarios, where the unmanaged memory allocation changes substantially during the lifetime of the managed object, you can call the AddMemoryPressure and RemoveMemoryPressure methods to communicate these incremental changes to the runtime.
>
> ### Caution
> You must ensure that you remove exactly the amount of pressure you add. Failing to do so can adversely affect the performance of the system in applications that run for long periods of time.

As far as I can tell, there's definitely an issue here. Although, I don't think it's got anything to do with the `Size + 1` part of the code. 

In my mind, the real issue is that the amount of memory pressure added in the call to `GC.AddMemoryPressure` is *not the same* as the call to `GC.RemoveMemoryPressure`. Would you agree?

I think the "correct" change is to make the call to `GC.AddMemoryPressure` the same as the call to `GC.RemoveMemoryPressure` like so:

```csharp
GC.AddMemoryPressure(SizeInBytes*Size);
```

What this is essentially telling the garbage collector is that we have pre-allocated that much memory to hold the particles. It doesn't actually matter how many particles we end up spawning, just that we have enough memory for all of them up front.

When the particle buffer is disposed, we tell the garbage collector that we've now released this unmanaged memory. 

From what I understand, the whole point of these methods is to make the garbage collector "behave nicely" while your game is running. Ideally, for the best possible performance we want to avoid doing garbage collections during rendering cycles and have them kick in during loading screens. This can be quite difficult to control in a memory managed run-time, but in any case we should be doing our best.

As for the actually affect all of this will have on your game, it's difficult to say. There's so many factors involved like how many particle buffer's you're allocating vs what platform you're running on vs what else your game is doing. 

Anyway, would you like to make this change and submit a pull request?

**Comment by luisfb** [comment]
The correct allocated memory value can be obtained from `Particle.SizeInBytes * Size`.
But the property SizeInBytes returns `Particle.SizeInBytes * (Size+1)`
I tried to remove that +1 but when I removed it I had infinite loops in the ParticleIterator because the HasNext returned always true.
This is being done for the property HasNext of the Iterator to return the right value when the end of the buffer is reached (after calling the method Next())
The end of the buffer is detected by evaluating the property BufferEnd (which uses Size+1). In fact, BufferEnd was the only thing I could detect that actually uses that (so far...).
So I Changed the code a bit.
I altered the BufferEnd property to use the +1 only on itself.
I also removed the +1 from prop SizeInBytes so it can return the right memory value.
I ran some unit tests on the iterator and it worked fine.
BUT... when I ran the real test... I mean, when I put this new code to run on my game... the particle effects in my game got messed up...
Then I discovered that the reason was because the method  Marshal.AllocHGlobal(SizeInBytes) also has to have the Size+1.

Conclusion:

Marshal.AllocHGlobal -> allocates the memory
GC.AddMemoryPressure -> tells garbage collector not to worry about the memory allocated
Marshal.FreeHGlobal -> Frees up the memory previously allocated.
GC.RemoveMemoryPressure -> tells GC the memory now is free.

Considering the above, all four methods should receive as parameter the same value of memory.
And since Marshal.AllocHGlobal needs the Size+1 to work properly, I am keeping the Size+1 and using that for all the methods.
So basically, what changes is the Dispose method:

From:
GC.RemoveMemoryPressure(Particle.SizeInBytes*Size);

To:
GC.RemoveMemoryPressure(SizeInBytes);

Pull Request:
#581

**Comment by craftworkgames** [comment]
Top work my friend. Thanks for looking into this in detail.
